### PR TITLE
Add async actions and background sound

### PR DESCRIPTION
**Pull request by syoung-smallwisdom**
Note: I discovered that with iOS 11, it doesn't play voice synthesizer if you lock the screen (or according to google if you background the app). Neat! It also won't play beeps and vibrations using timer delay. 

I am currently working around this by using GPS updates to fire the timer and using a vibrating alarm to indicate that you have ended your run. We should look into playing an mp3 or something to keep things active.

### CODE REVIEW

File: ResearchSuite/ResearchSuite/RSDAsyncActionController.swift [hunk 1]
```diff
@@ -66,17 +79,9 @@ public protocol RSDAsyncActionController : class {
...
-     A required initializer.
-     
-     @param configuration       The configuration for this controller.
-     @param outputDirectory     An output directory for where to save file results (if applicable)
-     */
-    init(configuration: RSDAsyncActionConfiguration, outputDirectory: URL?)
-    
-    /**
-     Start the asyncronous action with the given completion handler.
+     Start the asyncronous action with the given completion handler. Note: The handler may be called on a background thread.
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDAsyncActionController.swift]: asyncronous --> asynchronous

File: ResearchSuite/ResearchSuite/RSDSampleRecorder.swift [hunk 2]
```diff
@@ -0,0 +1,378 @@
+//
+//  RSDRecorder.swift
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDSampleRecorder.swift]: should match file name

File: ResearchSuite/ResearchSuite/RSDSampleRecorder.swift [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+            return collectionResult.inputResults.first
+        } else {
+            return collectionResult
+        }
+    }
+    
+    // MARK: State management
+    
+    public let isSimulator: Bool = {
+        #if arch(i386) || arch(x86_64)
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDSampleRecorder.swift]: this wouldn't work if building for MacOS, which is not an immediate priority but something to consider for a framework that is meant to be usable across various Apple OSes.

File: ResearchSuite/ResearchSuite/RSDSampleRecorder.swift [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+    
+    public final func stop(_ completion: RSDAsyncActionCompletionHandler?) {
+        self.loggerQueue.async {
+            var error: Error?
+            do {
+                try self._stopLogger()
+            } catch let err {
+                error = err
+            }
+            DispatchQueue.global().async {
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDSampleRecorder.swift]: I believe all the global queues (except main) are concurrent--could it be a problem if, say, stopRecorder() were to be called before startRecorder() had finished doing its thing?

File: ResearchSuite/ResearchSuite/RSDTaskController.swift [hunk 5]
```diff
@@ -325,40 +340,83 @@ extension RSDTaskController {
...
+                !taskPath.task!.stepNavigator.hasStep(after: nextStep!, with: taskPath.result))
+        
+        if !hasPreviousEarlyExit, let stopStep = previousStep,
+            let controllers = _asyncActionsToStop(after: stopStep, isTaskComplete: isTaskComplete) {
+            // If there are action controllers to stop then do that before continuing
+            // and call this method again after they have been stopped. Use the hasPreviousEarlyExit
+            // flag to indicate that this is the second go-around and don't check the controllers the
+            // second time.
+            DispatchQueue.global().async {
+                self.stopAsyncActions(for: controllers, completion: { [weak self] in
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTaskController.swift]: again, I'm pretty sure the global queues other than main are all concurrent, so stopAsyncActions() would need to be reentrant.

File: ResearchSuite/ResearchSuite/RSDTaskController.swift [hunk 6]
```diff
@@ -325,40 +340,83 @@ extension RSDTaskController {
...
+        let isFirstTaskStep = self.taskPath.currentStep == nil && self.taskPath.parentPath == nil
+        let isFirstSubtaskStep = self.taskPath.currentStep == nil
+        
+        if !hasPreviousEarlyExit, let asyncActions = _asyncActionsToStart(at: step, isFirstStep: isFirstSubtaskStep) {
+            // If there are action controllers to start then do that before continuing
+            // and call this method again after they have been started. Use the hasPreviousEarlyExit
+            // flag to indicate that this is the second go-around and don't check the controllers the
+            // second time.
+            DispatchQueue.global().async {
+                self.startAsyncActions(with: asyncActions) { [weak self] in
```
Comment (Erin-Mounts) [inline ResearchSuite/ResearchSuite/RSDTaskController.swift]: ditto for startAsyncActions()

File: ResearchSuiteUI/ResearchSuiteUI/iOS/RSDStepViewController.swift [hunk 7]
```diff
@@ -504,6 +514,9 @@ open class RSDStepViewController : UIViewController, RSDStepViewControllerProtoc
         }
     }
     
+    /**
+     Perform the stop commands. The
```
Comment (Erin-Mounts) [inline ResearchSuiteUI/ResearchSuiteUI/iOS/RSDStepViewController.swift]: The...?

**Review by Erin-Mounts — APPROVED** [review]
Couple of nits, couple of questions about some instances of using global (concurrent) dispatch queues, otherwise looks good, merge when ready

File: ResearchSuite/ResearchSuite/RSDSampleRecorder.swift (hunk 4 above)
Comment (syoung-smallwisdom) [reply ResearchSuite/ResearchSuite/RSDSampleRecorder.swift]: I couldn't really think of a good way to allow any subclass implementations to have the flexibility that they *might* need without the risk of enough rope to hang themselves? I guess I could call on the main thread? That's where I end up with the location recorder anyway.

File: ResearchSuite/ResearchSuite/RSDSampleRecorder.swift (hunk 3 above)
Comment (syoung-smallwisdom) [reply ResearchSuite/ResearchSuite/RSDSampleRecorder.swift]: good catch. thanks

File: ResearchSuite/ResearchSuite/RSDTaskController.swift (hunk 5 above)
Comment (syoung-smallwisdom) [reply ResearchSuite/ResearchSuite/RSDTaskController.swift]: That's the point. Using the global queue says "go put this on a concurrent background thread" but let other things happen.  In this case it's redundant, but I was trying to consider the case where someone is implementing their own TaskViewController. I suppose if they are doing that then they should know enough not to block the main thread. :)

**Pull request merged**